### PR TITLE
Error while signing on using Discord if no avatar #1649

### PR DESCRIPTION
**Pull request by venkatesh2k3**
**Description**

Changes Made:
Enhanced handling of the avatar field in SignOnControllerDiscord.java to ensure robust behavior when the avatar is missing or not a string.

**Details:**

I added validation to check for the avatar field in the JSON response, implemented logic to handle empty or non-string avatar fields, and introduced a default avatar URL fallback mechanism.

**Impact:**

This update ensures the application handles sign-on scenarios without Discord avatars gracefully and consistently, providing a stable user experience.

**Testing:**

Tested locally to verify functionality across different scenarios (avatar present, empty avatar).

**Note:**

The default avatar URL (https://e7.pngegg.com/pngimages/84/165/png-clipart-united-states-avatar-organization-information-user-avatar-service-computer-wallpaper-thumbnail.png) used in this implementation can be replaced with an actual URL pointing to a suitable default image for the application.

### CODE REVIEW

File: src/main/java/ai/elimu/web/SignOnControllerDiscord.java [hunk 1]
```diff
@@ -132,10 +132,25 @@ public String handleCallback(HttpServletRequest request) throws IOException, Int
                 String id = String.valueOf(idAsLong);
                 contributor.setProviderIdDiscord(id);
             }
+            private static final String DEFAULT_AVATAR_URL = "https://e7.pngegg.com/pngimages/84/165/"
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/web/SignOnControllerDiscord.java]: @venkatesh2k3 The `imageUrl` is not a required field: https://github.com/elimu-ai/webapp/blob/main/src/main/java/ai/elimu/model/contributor/Contributor.java#L57

Therefore, we don't need to fallback to a default image. Instead, it would be simpler (less code) to just leave it empty.

**Review by jo-elimu — CHANGES_REQUESTED** [review]
See if you can solve it with less code.

File: src/main/java/ai/elimu/web/SignOnControllerDiscord.java [hunk 2]
```diff
@@ -132,10 +132,20 @@ public String handleCallback(HttpServletRequest request) throws IOException, Int
...
+                    String avatar = jsonObject.getString("avatar");
+                    if (!avatar.isEmpty()) {
+                        String uriAvatar = "https://cdn.discordapp.com/avatars/" + jsonObject.getLong("id") + "/" + avatar + ".png";
+                        logger.info(uriAvatar);
+                        contributor.setImageUrl(uriAvatar);
+                    } else {
+                      logger.info("Avatar field is empty.");
+                    } 
+                } else {
+                    logger.warn("Avatar field is not a string: " + jsonObject.get("avatar"));
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/web/SignOnControllerDiscord.java:147]: @venkatesh2k3 What does the JSON response look like when this scenario ("Avatar field is not a string") occurs?

**Review by jo-elimu — CHANGES_REQUESTED** [review]
@venkatesh2k3 It should be possible to fix this bug with one or two lines of code, to keep the amount of code to a minimum.

Do you think it would work to replace `jsonObject.has("avatar")` with something like `jsonObject.has("avatar") && !StringUtils.isBlank(jsonObject.get("avatar"))`?

I don't know what the JSON response looks like when the bug occurs, but maybe this additional check would work?

Or maybe `jsonObject.has("avatar") && (jsonObject.get("avatar") instanceof String)`?

**Comment by venkatesh2k3** [comment]
> @venkatesh2k3 It should be possible to fix this bug with one or two lines of code, to keep the amount of code to a minimum.
> 
> Do you think it would work to replace `jsonObject.has("avatar")` with something like `jsonObject.has("avatar") && !StringUtils.isBlank(jsonObject.get("avatar"))`?
> 
> I don't know what the JSON response looks like when the bug occurs, but maybe this additional check would work?
> 
> Or maybe `jsonObject.has("avatar") && (jsonObject.get("avatar") instanceof String)`?

Yeah it will optimize code

**Review by jo-elimu — APPROVED** [review]

File: src/main/java/ai/elimu/web/SignOnControllerDiscord.java [hunk 3]
```diff
@@ -132,11 +132,13 @@ public String handleCallback(HttpServletRequest request) throws IOException, Int
                 String id = String.valueOf(idAsLong);
                 contributor.setProviderIdDiscord(id);
             }
-            if (jsonObject.has("avatar")) {
-                String uriAvatar = "https://cdn.discordapp.com/avatars/" + jsonObject.getLong("id") + "/" + jsonObject.getString("avatar") + ".png";
-                logger.info(uriAvatar);
-                contributor.setImageUrl(uriAvatar);
+
+            if (!jsonObject.isNull("avatar")) {
+            String uriAvatar = "https://cdn.discordapp.com/avatars/" + jsonObject.getLong("id") + "/" + jsonObject.getString("avatar") + ".png";
```
Comment (jo-elimu) [inline src/main/java/ai/elimu/web/SignOnControllerDiscord.java:137]: Looks like indentation is missing here.

**Review by jo-elimu — CHANGES_REQUESTED** [review]

**Review by jo-elimu — CHANGES_REQUESTED** [review]
Still no indentation. Maybe a tabs vs spaces problem? I think the default in this project is using 4 spaces instead of tabs.

image

**Review by jo-elimu — CHANGES_REQUESTED** [review]
There seems to be a compilation error: https://github.com/elimu-ai/webapp/actions/runs/9758324950/job/26932626216?pr=1650

File: src/main/java/ai/elimu/web/SignOnControllerDiscord.java [hunk 4]
```diff
@@ -132,9 +133,9 @@
                 String id = String.valueOf(idAsLong);
                 contributor.setProviderIdDiscord(id);
             }
-            if (jsonObject.has("avatar")) {
+            if (!jsonObject.isNull("avatar")) {
                 String uriAvatar = "https://cdn.discordapp.com/avatars/" + jsonObject.getLong("id") + "/" + jsonObject.getString("avatar") + ".png";
-                logger.info(uriAvatar);
+                logger.info("Avatar URL: " + uriAvatar);
```
[the remark this answers is not in the document]
Comment (jo-elimu) [reply src/main/java/ai/elimu/web/SignOnControllerDiscord.java:137]: @coderabbitai Yes, please generate the unit testing code.

[the remark this answers is not in the document]
Comment (jo-elimu) [reply src/main/java/ai/elimu/web/SignOnControllerDiscord.java:137]: @coderabbitai Can you add a unit test for testing `"avatar": null` in the response body?

**Review by jo-elimu — APPROVED** [review]

**Review by jo-elimu — APPROVED** [review]
Ready to merge 👍🏽
